From 002841da48791390e733e3fd235453832596d4ae Mon Sep 17 00:00:00 2001
From: Petar Kirov <petar.p.kirov@gmail.com>
Date: Thu, 15 Oct 2026 12:00:45 +0300
Subject: [PATCH] [SPIRV] Build logical access chains with 32-bit literal
 indices

`buildLogicalAccessChainFromGEP` turns a byte-addressing GEP back into an
access chain, and gave its literal indices an `i64` type while the leading
index is `i32`. A module with such an access chain needs the `Int64`
capability for nothing: a struct field read through an optimised GEP into
a uniform was enough, and a GLSL ES 1.00 cross-compile of the module then
fails ("64-bit integers not supported in ES profile before version 310").
Build them as `i32`, as the first index and the dynamic path's indices are.
---
 llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp |  5 ++-
 .../SPIRV/hlsl-resources/cbuffer-struct.ll    | 27 +++++++---------
 .../getelementptr-byte-addressing-array.ll    |  1 -
 ...ructured-buffer-access-constant-index-1.ll |  7 ++--
 ...ructured-buffer-access-constant-index-2.ll |  5 ++-
 .../pointers/structured-buffer-access.ll      | 32 +++++++++----------
 .../structured-buffer-vector-access.ll        | 22 ++++++-------
 7 files changed, 48 insertions(+), 51 deletions(-)

diff --git a/llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp b/llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp
index defc1e9322..ff7673c4e8 100644
--- a/llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp
+++ b/llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp
@@ -885,8 +885,11 @@ Instruction *SPIRVEmitIntrinsicsImpl::buildLogicalAccessChainFromGEP(
   walkLogicalAccessChain(
       GEP,
       [&Indices, &B](Type *EltType, uint64_t Index) {
+        // 32-bit, as the other indices are: a struct member index must be
+        // a 32-bit constant, and an `Int64` one needs the `Int64`
+        // capability that GLSL ES 1.00 cannot express.
         Indices.push_back(
-            ConstantInt::get(B.getInt64Ty(), Index, /* Signed= */ false));
+            ConstantInt::get(B.getInt32Ty(), Index, /* Signed= */ false));
       },
       [&Indices, &B, &DL, this](Type *EltType, Value *Offset,
                                 uint64_t Multiplier) {
diff --git a/llvm/test/CodeGen/SPIRV/hlsl-resources/cbuffer-struct.ll b/llvm/test/CodeGen/SPIRV/hlsl-resources/cbuffer-struct.ll
index 493e6dc4ac..7fe0666749 100644
--- a/llvm/test/CodeGen/SPIRV/hlsl-resources/cbuffer-struct.ll
+++ b/llvm/test/CodeGen/SPIRV/hlsl-resources/cbuffer-struct.ll
@@ -18,7 +18,6 @@
 
 ; CHECK-DAG: %[[STRUCT_CBUFFER:[0-9]+]] = OpTypeStruct %[[STRUCT_INNER]]
 ; CHECK-DAG: %[[PTR_CBUFFER:[0-9]+]] = OpTypePointer Uniform %[[STRUCT_CBUFFER]]
-; CHECK-DAG: %[[INT64:[0-9]+]] = OpTypeInt 64 0
 
 ; CHECK-DAG: OpMemberDecorate %[[STRUCT_CBUFFER]] 0 Offset 0
 ; CHECK-DAG: OpDecorate %[[STRUCT_CBUFFER]] Block
@@ -33,10 +32,8 @@
 ; CHECK-DAG: OpMemberDecorate %[[STRUCT_MATRIX]] 3 Offset 48
 
 ; CHECK-DAG: %[[ONE:[0-9]+]] = OpConstant %[[INT]] 1{{$}}
-; CHECK-DAG: %[[ZERO_64:[0-9]+]] = OpConstant %[[INT64]] 0{{$}}
-; CHECK-DAG: %[[ONE_64:[0-9]+]] = OpConstant %[[INT64]] 1{{$}}
-; CHECK-DAG: %[[TWO_64:[0-9]+]] = OpConstant %[[INT64]] 2{{$}}
-; CHECK-DAG: %[[THREE_64:[0-9]+]] = OpConstant %[[INT64]] 3{{$}}
+; CHECK-DAG: %[[TWO:[0-9]+]] = OpConstant %[[INT]] 2{{$}}
+; CHECK-DAG: %[[THREE:[0-9]+]] = OpConstant %[[INT]] 3{{$}}
 
 ; CHECK: %[[CBUFFER:[0-9]+]] = OpVariable %[[PTR_CBUFFER]] Uniform
 
@@ -73,11 +70,11 @@ entry:
 ; CHECK: %[[VAL_M0_V0:[0-9]+]] = OpLoad %[[VEC4]] %[[PTR_M0_V0]]
   %4 = load <4 x float>, ptr addrspace(12) @transforms, align 16
 
-; CHECK: %[[PTR_M0_V1:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[ZERO_64]] %[[ONE_64]]
+; CHECK: %[[PTR_M0_V1:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[ZERO]] %[[ONE]]
 ; CHECK: %[[VAL_M0_V1:[0-9]+]] = OpLoad %[[VEC4]] %[[PTR_M0_V1]]
   %5 = load <4 x float>, ptr addrspace(12) getelementptr inbounds nuw (i8, ptr addrspace(12) @transforms, i64 16), align 16
 
-; CHECK: %[[PTR_M0_V3:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[ZERO_64]] %[[THREE_64]]
+; CHECK: %[[PTR_M0_V3:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[ZERO]] %[[THREE]]
 ; CHECK: %[[VAL_M0_V3:[0-9]+]] = OpLoad %[[VEC4]] %[[PTR_M0_V3]]
   %6 = load <4 x float>, ptr addrspace(12) getelementptr inbounds nuw (i8, ptr addrspace(12) @transforms, i64 48), align 16
 
@@ -87,17 +84,17 @@ entry:
   %mul3.i20.i = fmul reassoc nnan ninf nsz arcp afn <4 x float> %splat.splat2.i19.i, %5
   %8 = tail call reassoc nnan ninf nsz arcp afn <4 x float> @llvm.fmuladd.v4f32(<4 x float> %splat.splat.i18.i, <4 x float> nofpclass(nan inf) %4, <4 x float> %mul3.i20.i)
   %9 = fadd reassoc nnan ninf nsz arcp afn <4 x float> %8, %6
-; CHECK: %[[PTR_M1:[0-9]+]] = OpInBoundsAccessChain %[[PTR_MATRIX]] %[[PTR_STRUCT]] %[[ONE_64]]
+; CHECK: %[[PTR_M1:[0-9]+]] = OpInBoundsAccessChain %[[PTR_MATRIX]] %[[PTR_STRUCT]] %[[ONE]]
 ; CHECK: %[[PTR_M1_V0:[0-9]+]] = OpAccessChain %[[PTR_VEC4]] %[[PTR_M1]] %[[ZERO]]
 ; CHECK: %[[VAL_M1_V0:[0-9]+]] = OpLoad %[[VEC4]] %[[PTR_M1_V0]]
   %10 = load <4 x float>, ptr addrspace(12) getelementptr inbounds nuw (i8, ptr addrspace(12) @transforms, i64 64), align 16
-; CHECK: %[[PTR_M1_V1:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[ONE_64]] %[[ONE_64]]
+; CHECK: %[[PTR_M1_V1:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[ONE]] %[[ONE]]
 ; CHECK: %[[VAL_M1_V1:[0-9]+]] = OpLoad %[[VEC4]] %[[PTR_M1_V1]]
   %11 = load <4 x float>, ptr addrspace(12) getelementptr inbounds nuw (i8, ptr addrspace(12) @transforms, i64 80), align 16
-; CHECK: %[[PTR_M1_V2:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[ONE_64]] %[[TWO_64]]
+; CHECK: %[[PTR_M1_V2:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[ONE]] %[[TWO]]
 ; CHECK: %[[VAL_M1_V2:[0-9]+]] = OpLoad %[[VEC4]] %[[PTR_M1_V2]]
   %12 = load <4 x float>, ptr addrspace(12) getelementptr inbounds nuw (i8, ptr addrspace(12) @transforms, i64 96), align 16
-; CHECK: %[[PTR_M1_V3:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[ONE_64]] %[[THREE_64]]
+; CHECK: %[[PTR_M1_V3:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[ONE]] %[[THREE]]
 ; CHECK: %[[VAL_M1_V3:[0-9]+]] = OpLoad %[[VEC4]] %[[PTR_M1_V3]]
   %13 = load <4 x float>, ptr addrspace(12) getelementptr inbounds nuw (i8, ptr addrspace(12) @transforms, i64 112), align 16
   %splat.splat.i13.i = shufflevector <4 x float> %9, <4 x float> poison, <4 x i32> zeroinitializer
@@ -108,17 +105,17 @@ entry:
   %15 = tail call reassoc nnan ninf nsz arcp afn <4 x float> @llvm.fmuladd.v4f32(<4 x float> %splat.splat5.i16.i, <4 x float> nofpclass(nan inf) %12, <4 x float> %14)
   %splat.splat7.i17.i = shufflevector <4 x float> %9, <4 x float> poison, <4 x i32> <i32 3, i32 3, i32 3, i32 3>
   %16 = tail call reassoc nnan ninf nsz arcp afn noundef <4 x float> @llvm.fmuladd.v4f32(<4 x float> %splat.splat7.i17.i, <4 x float> nofpclass(nan inf) %13, <4 x float> %15)
-; CHECK: %[[PTR_M2:[0-9]+]] = OpInBoundsAccessChain %[[PTR_MATRIX]] %[[PTR_STRUCT]] %[[TWO_64]]
+; CHECK: %[[PTR_M2:[0-9]+]] = OpInBoundsAccessChain %[[PTR_MATRIX]] %[[PTR_STRUCT]] %[[TWO]]
 ; CHECK: %[[PTR_M2_V0:[0-9]+]] = OpAccessChain %[[PTR_VEC4]] %[[PTR_M2]] %[[ZERO]]
 ; CHECK: %[[VAL_M2_V0:[0-9]+]] = OpLoad %[[VEC4]] %[[PTR_M2_V0]]
   %17 = load <4 x float>, ptr addrspace(12) getelementptr inbounds nuw (i8, ptr addrspace(12) @transforms, i64 128), align 16
-; CHECK: %[[PTR_M2_V1:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[TWO_64]] %[[ONE_64]]
+; CHECK: %[[PTR_M2_V1:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[TWO]] %[[ONE]]
 ; CHECK: %[[VAL_M2_V1:[0-9]+]] = OpLoad %[[VEC4]] %[[PTR_M2_V1]]
   %18 = load <4 x float>, ptr addrspace(12) getelementptr inbounds nuw (i8, ptr addrspace(12) @transforms, i64 144), align 16
-; CHECK: %[[PTR_M2_V2:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[TWO_64]] %[[TWO_64]]
+; CHECK: %[[PTR_M2_V2:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[TWO]] %[[TWO]]
 ; CHECK: %[[VAL_M2_V2:[0-9]+]] = OpLoad %[[VEC4]] %[[PTR_M2_V2]]
   %19 = load <4 x float>, ptr addrspace(12) getelementptr inbounds nuw (i8, ptr addrspace(12) @transforms, i64 160), align 16
-; CHECK: %[[PTR_M2_V3:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[TWO_64]] %[[THREE_64]]
+; CHECK: %[[PTR_M2_V3:[0-9]+]] = OpInBoundsAccessChain %[[PTR_VEC4]] %[[PTR_STRUCT]] %[[TWO]] %[[THREE]]
 ; CHECK: %[[VAL_M2_V3:[0-9]+]] = OpLoad %[[VEC4]] %[[PTR_M2_V3]]
   %20 = load <4 x float>, ptr addrspace(12) getelementptr inbounds nuw (i8, ptr addrspace(12) @transforms, i64 176), align 16
   %splat.splat.i.i = shufflevector <4 x float> %16, <4 x float> poison, <4 x i32> zeroinitializer
diff --git a/llvm/test/CodeGen/SPIRV/pointers/getelementptr-byte-addressing-array.ll b/llvm/test/CodeGen/SPIRV/pointers/getelementptr-byte-addressing-array.ll
index 4368323e0e..d0dcc8b2e4 100644
--- a/llvm/test/CodeGen/SPIRV/pointers/getelementptr-byte-addressing-array.ll
+++ b/llvm/test/CodeGen/SPIRV/pointers/getelementptr-byte-addressing-array.ll
@@ -2,7 +2,6 @@
 ; RUN: %if spirv-tools %{ llc -O0 -mtriple=spirv1.6-unknown-vulkan1.3-compute %s -o - -filetype=obj | spirv-val --target-env vulkan1.3 %}
 
 ; CHECK-DAG: %[[#int32:]] = OpTypeInt 32 0
-; CHECK-DAG: %[[#int64:]] = OpTypeInt 64 0
 ; CHECK-DAG: %[[#float:]] = OpTypeFloat 32
 ; CHECK-DAG: %[[#vec2:]] = OpTypeVector %[[#float]] 2
 ; CHECK-DAG: %[[#arr3:]] = OpTypeArray %[[#vec2]] %[[#]]
diff --git a/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access-constant-index-1.ll b/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access-constant-index-1.ll
index fce0f9798a..7f36e83a9e 100644
--- a/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access-constant-index-1.ll
+++ b/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access-constant-index-1.ll
@@ -11,13 +11,12 @@ entry:
   %0 = tail call target("spirv.VulkanBuffer", [0 x %struct.S1], 12, 0) @llvm.spv.resource.handlefrombinding.tspirv.VulkanBuffer_a0s_struct.S1s_12_0t(i32 0, i32 1, i32 1, i32 0, ptr nonnull @.str)
   %3 = tail call noundef align 1 dereferenceable(192) ptr addrspace(11) @llvm.spv.resource.getpointer.p11.tspirv.VulkanBuffer_a0s_struct.S1s_12_0t(target("spirv.VulkanBuffer", [0 x %struct.S1], 12, 0) %0, i32 0)
 
-; CHECK-DAG:  %[[#ulong:]] = OpTypeInt 64 0
-; CHECK-DAG:  %[[#ulong_1:]] = OpConstant %[[#ulong]] 1
-; CHECK-DAG:  %[[#ulong_3:]] = OpConstant %[[#ulong]] 3
 
 ; CHECK-DAG:  %[[#uint:]] = OpTypeInt 32 0
 ; CHECK-DAG:  %[[#uint_0:]] = OpConstant %[[#uint]] 0
 ; CHECK-DAG:  %[[#uint_10:]] = OpConstant %[[#uint]] 10
+; CHECK-DAG:  %[[#uint_1:]] = OpConstant %[[#uint]] 1
+; CHECK-DAG:  %[[#uint_3:]] = OpConstant %[[#uint]] 3
 
 ; CHECK-DAG:  %[[#float:]] = OpTypeFloat 32
 ; CHECK-DAG:  %[[#v4f:]] = OpTypeVector %[[#float]] 4
@@ -27,7 +26,7 @@ entry:
 ; CHECK-DAG:  %[[#sb_v4f:]] = OpTypePointer StorageBuffer %[[#v4f]]
 
 ; CHECK:      %[[#tmp:]] = OpAccessChain %[[#sb_S1]] %[[#]] %[[#uint_0]] %[[#uint_0]]
-; CHECK:      %[[#ptr:]] = OpInBoundsAccessChain %[[#sb_v4f]] %[[#tmp]] %[[#ulong_1]] %[[#ulong_3]]
+; CHECK:      %[[#ptr:]] = OpInBoundsAccessChain %[[#sb_v4f]] %[[#tmp]] %[[#uint_1]] %[[#uint_3]]
 ; This rewritten GEP combined all constant indices into a single value.
 ; We should make sure the correct indices are retrieved.
   %arrayidx.i = getelementptr inbounds nuw i8, ptr addrspace(11) %3, i64 64
diff --git a/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access-constant-index-2.ll b/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access-constant-index-2.ll
index c676007e80..cc1a3b3aad 100644
--- a/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access-constant-index-2.ll
+++ b/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access-constant-index-2.ll
@@ -11,13 +11,12 @@ entry:
   %0 = tail call target("spirv.VulkanBuffer", [0 x %struct.S1], 12, 0) @llvm.spv.resource.handlefrombinding.tspirv.VulkanBuffer_a0s_struct.S1s_12_0t(i32 0, i32 1, i32 1, i32 0, ptr nonnull @.str)
   %3 = tail call noundef align 1 dereferenceable(192) ptr addrspace(11) @llvm.spv.resource.getpointer.p11.tspirv.VulkanBuffer_a0s_struct.S1s_12_0t(target("spirv.VulkanBuffer", [0 x %struct.S1], 12, 0) %0, i32 0)
 
-; CHECK-DAG:  %[[#ulong:]] = OpTypeInt 64 0
-; CHECK-DAG:  %[[#ulong_1:]] = OpConstant %[[#ulong]] 1
 
 ; CHECK-DAG:  %[[#uint:]] = OpTypeInt 32 0
 ; CHECK-DAG:  %[[#uint_0:]] = OpConstant %[[#uint]] 0
 ; CHECK-DAG:  %[[#uint_10:]] = OpConstant %[[#uint]] 10
 ; CHECK-DAG:  %[[#uint_16:]] = OpConstant %[[#uint]] 16
+; CHECK-DAG:  %[[#uint_1:]] = OpConstant %[[#uint]] 1
 
 ; CHECK-DAG:  %[[#float:]] = OpTypeFloat 32
 ; CHECK-DAG:  %[[#v4f:]] = OpTypeVector %[[#float]] 4
@@ -28,7 +27,7 @@ entry:
 ; CHECK-DAG:  %[[#sb_v4f:]] = OpTypePointer StorageBuffer %[[#v4f]]
 
 ; CHECK:      %[[#a:]] = OpAccessChain %[[#sb_S1]] %[[#]] %[[#uint_0]] %[[#uint_0]]
-; CHECK:      %[[#b:]] = OpInBoundsAccessChain %[[#sb_arr_v4f]] %[[#a]] %[[#ulong_1]]
+; CHECK:      %[[#b:]] = OpInBoundsAccessChain %[[#sb_arr_v4f]] %[[#a]] %[[#uint_1]]
   %4 = getelementptr inbounds nuw i8, ptr addrspace(11) %3, i64 16
 
 ; CHECK:      %[[#offset:]] = OpIMul %[[#]] %[[#]] %[[#uint_16]]
diff --git a/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access.ll b/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access.ll
index 8400db5b74..6de2384dd2 100644
--- a/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access.ll
+++ b/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-access.ll
@@ -28,23 +28,23 @@
 
 define void @main() local_unnamed_addr #0 {
 ; CHECK-LABEL: main
-; CHECK:       %42 = OpFunction %2 None %3 ; -- Begin function main
+; CHECK:       %41 = OpFunction %2 None %3 ; -- Begin function main
 ; CHECK-NEXT:    %1 = OpLabel
-; CHECK-NEXT:    %43 = OpVariable %27 Function %37
-; CHECK-NEXT:    %44 = OpVariable %26 Function %38
-; CHECK-NEXT:    %45 = OpCopyObject %19 %39
-; CHECK-NEXT:    %46 = OpCopyObject %16 %40
-; CHECK-NEXT:    %47 = OpLoad %4 %41
-; CHECK-NEXT:    %48 = OpAccessChain %13 %45 %28 %47
-; CHECK-NEXT:    %49 = OpInBoundsAccessChain %9 %48 %30
-; CHECK-NEXT:    %50 = OpLoad %8 %49
-; CHECK-NEXT:    %51 = OpAccessChain %11 %46 %28 %47
-; CHECK-NEXT:    %52 = OpInBoundsAccessChain %9 %51 %28
-; CHECK-NEXT:    OpStore %52 %50
-; CHECK-NEXT:    %53 = OpAccessChain %6 %48 %28
-; CHECK-NEXT:    %54 = OpLoad %5 %53
-; CHECK-NEXT:    %55 = OpInBoundsAccessChain %6 %51 %30
-; CHECK-NEXT:    OpStore %55 %54
+; CHECK-NEXT:    %42 = OpVariable %26 Function %36
+; CHECK-NEXT:    %43 = OpVariable %25 Function %37
+; CHECK-NEXT:    %44 = OpCopyObject %19 %38
+; CHECK-NEXT:    %45 = OpCopyObject %16 %39
+; CHECK-NEXT:    %46 = OpLoad %4 %40
+; CHECK-NEXT:    %47 = OpAccessChain %13 %44 %27 %46
+; CHECK-NEXT:    %48 = OpInBoundsAccessChain %9 %47 %29
+; CHECK-NEXT:    %49 = OpLoad %8 %48
+; CHECK-NEXT:    %50 = OpAccessChain %11 %45 %27 %46
+; CHECK-NEXT:    %51 = OpInBoundsAccessChain %9 %50 %27
+; CHECK-NEXT:    OpStore %51 %49
+; CHECK-NEXT:    %52 = OpAccessChain %6 %47 %27
+; CHECK-NEXT:    %53 = OpLoad %5 %52
+; CHECK-NEXT:    %54 = OpInBoundsAccessChain %6 %50 %29
+; CHECK-NEXT:    OpStore %54 %53
 ; CHECK-NEXT:    OpReturn
 ; CHECK-NEXT:    OpFunctionEnd
 entry:
diff --git a/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-vector-access.ll b/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-vector-access.ll
index d7681787ec..5ac00e6dc2 100644
--- a/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-vector-access.ll
+++ b/llvm/test/CodeGen/SPIRV/pointers/structured-buffer-vector-access.ll
@@ -15,18 +15,18 @@
 
 define void @main() local_unnamed_addr #0 {
 ; CHECK-LABEL: main
-; CHECK:       %33 = OpFunction %2 None %3 ; -- Begin function main
+; CHECK:       %32 = OpFunction %2 None %3 ; -- Begin function main
 ; CHECK-NEXT:    %1 = OpLabel
-; CHECK-NEXT:    %34 = OpVariable %20 Function %29
-; CHECK-NEXT:    %35 = OpVariable %19 Function %30
-; CHECK-NEXT:    %36 = OpCopyObject %12 %31
-; CHECK-NEXT:    %37 = OpCopyObject %10 %32
-; CHECK-NEXT:    %38 = OpAccessChain %7 %36 %21 %21
-; CHECK-NEXT:    %39 = OpLoad %6 %38
-; CHECK-NEXT:    %40 = OpCompositeExtract %4 %39 1
-; CHECK-NEXT:    %41 = OpAccessChain %7 %37 %21 %21
-; CHECK-NEXT:    %42 = OpInBoundsAccessChain %5 %41 %22
-; CHECK-NEXT:    OpStore %42 %40
+; CHECK-NEXT:    %33 = OpVariable %19 Function %28
+; CHECK-NEXT:    %34 = OpVariable %18 Function %29
+; CHECK-NEXT:    %35 = OpCopyObject %12 %30
+; CHECK-NEXT:    %36 = OpCopyObject %10 %31
+; CHECK-NEXT:    %37 = OpAccessChain %7 %35 %20 %20
+; CHECK-NEXT:    %38 = OpLoad %6 %37
+; CHECK-NEXT:    %39 = OpCompositeExtract %4 %38 1
+; CHECK-NEXT:    %40 = OpAccessChain %7 %36 %20 %20
+; CHECK-NEXT:    %41 = OpInBoundsAccessChain %5 %40 %21
+; CHECK-NEXT:    OpStore %41 %39
 ; CHECK-NEXT:    OpReturn
 ; CHECK-NEXT:    OpFunctionEnd
 entry:
-- 
2.51.2

